File: src/DelegateDecompiler.Tests/DecimalTests.cs
using System;
using System.Linq.Expressions;
using NUnit.Framework;

namespace DelegateDecompiler.Tests
{
    [TestFixture]
    public class DecimalTests : DecompilerTestsBase
    {
        [Test]
        public void Ceq()
        {
            Expression<Func<decimal, decimal, bool>> expected = (x, y) => x == y;
            Func<decimal, decimal, bool> compiled = (x, y) => x == y;
            Test(expected, compiled);
        }

        [Test]
        public void Cgt()
        {
            Expression<Func<decimal, decimal, bool>> expected = (x, y) => x > y;
            Func<decimal, decimal, bool> compiled = (x, y) => x > y;
            Test(expected, compiled);
        }

        [Test]
        public void Bgt()
        {
            Expression<Func<decimal, decimal, decimal>> expected = (x, y) => x > y ? x : y;
            Func<decimal, decimal, decimal> compiled = (x, y) => x > y ? x : y;
            Test(expected, compiled);
        }

        [Test]
        public void Cge()
        {
            Expression<Func<decimal, decimal, bool>> expected = (x, y) => x >= y;
            Func<decimal, decimal, bool> compiled = (x, y) => x >= y;
            Test(expected, compiled);
        }

        [Test]
        public void Bge()
        {
            Expression<Func<decimal, decimal, decimal>> expected = (x, y) => x >= y ? x : y;
            Func<decimal, decimal, decimal> compiled = (x, y) => x >= y ? x : y;
            Test(expected, compiled);
        }

        [Test]
        public void Clt()
        {
            Expression<Func<decimal, decimal, bool>> expected = (x, y) => x < y;
            Func<decimal, decimal, bool> compiled = (x, y) => x < y;
            Test(expected, compiled);
        }

        [Test]
        public void Blt()
        {
            Expression<Func<decimal, decimal, decimal>> expected = (x, y) => x < y ? x : y;
            Func<decimal, decimal, decimal> compiled = (x, y) => x < y ? x : y;
            Test(expected, compiled);
        }

        [Test]
        public void Cle()
        {
            Expression<Func<decimal, decimal, bool>> expected = (x, y) => x <= y;
            Func<decimal, decimal, bool> compiled = (x, y) => x <= y;
            Test(expected, compiled);
        }

        [Test]
        public void Ble()
        {
            Expression<Func<decimal, decimal, decimal>> expected = (x, y) => x <= y ? x : y;
            Func<decimal, decimal, decimal> compiled = (x, y) => x <= y ? x : y;
            Test(expected, compiled);
        }

        [Test]
        public void Bne()
        {
            Expression<Func<decimal, decimal, decimal>> expected = (x, y) => x != y ? x : y;
            Func<decimal, decimal, decimal> compiled = (x, y) => x != y ? x : y;
            Test(expected, compiled);
        }

        [Test]
        public void Beq()
        {
            Expression<Func<decimal, decimal, decimal>> expected = (x, y) => x == y ? x : y;
            Func<decimal, decimal, decimal> compiled = (x, y) => x == y ? x : y;
            Test(expected, compiled);
        }

        [Test]
        public void Mul()
        {
            Expression<Func<decimal, decimal, decimal>> expected = (x, y) => x * y;
            Func<decimal, decimal, decimal> compiled = (x, y) => x * y;
            Test(expected, compiled);
        }

        [Test]
        public void MulByInt()
        {
            Expression<Func<decimal, int, decimal>> expected = (x, y) => x * y;
            Func<decimal, int, decimal> compiled = (x, y) => x * y;
            Test(expected, compiled);
        }

        [Test]
        public void Div()
        {
            Expression<Func<decimal, decimal, decimal>> expected = (x, y) => x / y;
            Func<decimal, decimal, decimal> compiled = (x, y) => x / y;
            Test(expected, compiled);
        }

        [Test]
        public void DecimalCast()
        {
            Expression<Func<decimal>> expected1 = () => (decimal)1.21;
            Expression<Func<decimal>> expected2 = () => new decimal(121, 0, 0, false, 2);
            Func<decimal> compiled = () => (decimal)1.21;
            Test(expected1, expected2, compiled);
        }

        [Test]
        public void DecimalConstructor()
        {
            Expression<Func<decimal>> expected1 = () => 1.21M;
            Expression<Func<decimal>> expected2 = () => new decimal(121, 0, 0, false, 2);
            Func<decimal> compiled = () => 1.21M;
            Test(expected1, expected2, compiled);
        }
    }
}

File: src/DelegateDecompiler.Tests.VB/DecimalTests.vb
Imports System
Imports System.Linq.Expressions
Imports NUnit.Framework


<TestFixture>
Public Class DecimalTests
    Inherits DecompilerTestsBase

    <Test>
    Public Sub Ceq()
        Dim expected As Expression(Of Func(Of Decimal, Decimal, Boolean)) = Function(x As Decimal, y As Decimal) x = y
        Dim compiled As Func(Of Decimal, Decimal, Boolean) = Function(x As Decimal, y As Decimal) x = y
        Test(expected, compiled)
    End Sub

    <Test>
    Public Sub Cgt()
        Dim expected As Expression(Of Func(Of Decimal, Decimal, Boolean)) = Function(x As Decimal, y As Decimal) x > y
        Dim compiled As Func(Of Decimal, Decimal, Boolean) = Function(x As Decimal, y As Decimal) x > y
        Test(expected, compiled)
    End Sub

    <Test>
    Public Sub Bgt()
        Dim expected As Expression(Of Func(Of Decimal, Decimal, Decimal)) = Function(x As Decimal, y As Decimal) If((x > y), x, y)
        Dim compiled As Func(Of Decimal, Decimal, Decimal) = Function(x As Decimal, y As Decimal) If((x > y), x, y)
        Test(expected, compiled)
    End Sub

    <Test>
    Public Sub Cge()
        Dim expected As Expression(Of Func(Of Decimal, Decimal, Boolean)) = Function(x As Decimal, y As Decimal) x >= y
        Dim compiled As Func(Of Decimal, Decimal, Boolean) = Function(x As Decimal, y As Decimal) x >= y
        Test(expected, compiled)
    End Sub

    <Test>
    Public Sub Bge()
        Dim expected As Expression(Of Func(Of Decimal, Decimal, Decimal)) = Function(x As Decimal, y As Decimal) If((x >= y), x, y)
        Dim compiled As Func(Of Decimal, Decimal, Decimal) = Function(x As Decimal, y As Decimal) If((x >= y), x, y)
        Test(expected, compiled)
    End Sub

    <Test>
    Public Sub Clt()
        Dim expected As Expression(Of Func(Of Decimal, Decimal, Boolean)) = Function(x As Decimal, y As Decimal) x < y
        Dim compiled As Func(Of Decimal, Decimal, Boolean) = Function(x As Decimal, y As Decimal) x < y
        Test(expected, compiled)
    End Sub

    <Test>
    Public Sub Blt()
        Dim expected As Expression(Of Func(Of Decimal, Decimal, Decimal)) = Function(x As Decimal, y As Decimal) If((x < y), x, y)
        Dim compiled As Func(Of Decimal, Decimal, Decimal) = Function(x As Decimal, y As Decimal) If((x < y), x, y)
        Test(expected, compiled)
    End Sub

    <Test>
    Public Sub Cle()
        Dim expected As Expression(Of Func(Of Decimal, Decimal, Boolean)) = Function(x As Decimal, y As Decimal) x <= y
        Dim compiled As Func(Of Decimal, Decimal, Boolean) = Function(x As Decimal, y As Decimal) x <= y
        Test(expected, compiled)
    End Sub

    <Test>
    Public Sub Ble()
        Dim expected As Expression(Of Func(Of Decimal, Decimal, Decimal)) = Function(x As Decimal, y As Decimal) If((x <= y), x, y)
        Dim compiled As Func(Of Decimal, Decimal, Decimal) = Function(x As Decimal, y As Decimal) If((x <= y), x, y)
        Test(expected, compiled)
    End Sub

    <Test>
    Public Sub Bne()
        Dim expected As Expression(Of Func(Of Decimal, Decimal, Decimal)) = Function(x As Decimal, y As Decimal) If((x <> y), x, y)
        Dim compiled As Func(Of Decimal, Decimal, Decimal) = Function(x As Decimal, y As Decimal) If((x <> y), x, y)
        Test(expected, compiled)
    End Sub

    <Test>
    Public Sub Beq()
        Dim expected As Expression(Of Func(Of Decimal, Decimal, Decimal)) = Function(x As Decimal, y As Decimal) If((x = y), x, y)
        Dim compiled As Func(Of Decimal, Decimal, Decimal) = Function(x As Decimal, y As Decimal) If((x = y), x, y)
        Test(expected, compiled)
    End Sub

    <Test>
    Public Sub Mul()
        Dim expected As Expression(Of Func(Of Decimal, Decimal, Decimal)) = Function(x As Decimal, y As Decimal) x*y
        Dim compiled As Func(Of Decimal, Decimal, Decimal) = Function(x As Decimal, y As Decimal) x*y
        Test(expected, compiled)
    End Sub

    <Test>
    Public Sub MulByInt()
        Dim expected As Expression(Of Func(Of Decimal, Integer, Decimal)) = Function(x As Decimal, y As Integer) x*y
        Dim compiled As Func(Of Decimal, Integer, Decimal) = Function(x As Decimal, y As Integer) x*y
        Test(expected, compiled)
    End Sub

    <Test>
    Public Sub Div()
        Dim expected As Expression(Of Func(Of Decimal, Decimal, Decimal)) = Function(x As Decimal, y As Decimal) x/y
        Dim compiled As Func(Of Decimal, Decimal, Decimal) = Function(x As Decimal, y As Decimal) x/y
        Test(expected, compiled)
    End Sub

    <Test>
    Public Sub DecimalCast()
        Dim expected As Expression(Of Func(Of Decimal)) = Function() 1.21D
        Dim expected2 As Expression(Of Func(Of Decimal)) = Function() New Decimal(121, 0, 0, False, 2)
        Dim compiled As Func(Of Decimal) = Function() 1.21D
        Test(expected, expected2, compiled)
    End Sub

    <Test>
    Public Sub DecimalConstructor()
        Dim expected As Expression(Of Func(Of Decimal)) = Function() 1.21D
        Dim expected2 As Expression(Of Func(Of Decimal)) = Function() New Decimal(121, 0, 0, False, 2)
        Dim compiled As Func(Of Decimal) = Function() New Decimal(121, 0, 0, False, 2)
        Test(expected, expected2, compiled)
    End Sub
End Class

